### PR TITLE
Feat/integration test for one time verification

### PR DESCRIPTION
**Pull request by motemotech**
All changes to add integration test in one time verification

### CODE REVIEW

**Comment by motemotech** [comment]
I will close this pr and separate into small changes

File: circuits/scripts/build_dsc_4096_circuits.sh [hunk 1]
```diff
@@ -36,23 +36,23 @@ build_circuit() {
     echo $RAND_STR | yarn snarkjs zkey contribute build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}.zkey build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}_final.zkey
     yarn snarkjs zkey export verificationkey build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}_final.zkey build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}_vkey.json
 
-    yarn snarkjs zkey export solidityverifier build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}_final.zkey build/dsc/Verifier_${CIRCUIT_NAME}.sol
-    sed -i '' "s/Groth16Verifier/Verifier_${CIRCUIT_NAME}/g" build/dsc/Verifier_${CIRCUIT_NAME}.sol
-    cp build/dsc/Verifier_${CIRCUIT_NAME}.sol ../contracts/contracts/dsc/Verifier_${CIRCUIT_NAME}.sol
+    yarn snarkjs zkey export solidityverifier build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}_final.zkey build/dsc/${CIRCUIT_NAME}/Verifier_${CIRCUIT_NAME}.sol
+    sed -i '' "s/Groth16Verifier/Verifier_${CIRCUIT_NAME}/g" build/dsc/${CIRCUIT_NAME}/Verifier_${CIRCUIT_NAME}.sol
+    cp build/dsc/${CIRCUIT_NAME}/Verifier_${CIRCUIT_NAME}.sol ../contracts/contracts/verifiers/local/dsc/Verifier_${CIRCUIT_NAME}.sol
```
Comment (motemotech) [inline circuits/scripts/build_dsc_4096_circuits.sh:41]: solidityverifier is not correctly exported.
Just changed the path

File: circuits/scripts/build_dsc_4096_circuits.sh [hunk 2]
```diff
@@ -36,23 +36,23 @@ build_circuit() {
...
 
     echo -e "\033[32mBuild of $CIRCUIT_NAME completed in $(($(date +%s) - START_TIME)) seconds\033[0m"
 
     echo "file sizes:"
-    echo -e "\033[34mSize of ${CIRCUIT_NAME}.r1cs: $(wc -c <build/${CIRCUIT_NAME}.r1cs) bytes\033[0m"
-    echo -e "\033[34mSize of ${CIRCUIT_NAME}.wasm: $(wc -c <build/${CIRCUIT_NAME}_js/${CIRCUIT_NAME}.wasm) bytes\033[0m"
-    echo -e "\033[34mSize of ${CIRCUIT_NAME}_final.zkey: $(wc -c <build/${CIRCUIT_NAME}_final.zkey) bytes\033[0m"
+    echo -e "\033[34mSize of ${CIRCUIT_NAME}.r1cs: $(wc -c <build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}.r1cs) bytes\033[0m"
+    echo -e "\033[34mSize of ${CIRCUIT_NAME}.wasm: $(wc -c <build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}_js/${CIRCUIT_NAME}.wasm) bytes\033[0m"
+    echo -e "\033[34mSize of ${CIRCUIT_NAME}_final.zkey: $(wc -c <build/dsc/${CIRCUIT_NAME}/${CIRCUIT_NAME}_final.zkey) bytes\033[0m"
```
Comment (motemotech) [inline circuits/scripts/build_dsc_4096_circuits.sh:49]: path was wrong.
Changed to correct one

File: circuits/scripts/download_circuits_from_aws.sh [hunk 3]
```diff
@@ -42,6 +42,9 @@ download_and_compile_circuit() {
                 --O1 --wasm \
                 --output build/fromAWS
 
+            echo "Exporting vkey: $circuit_name"
+            snarkjs zkey export verificationkey build/fromAWS/${circuit_name}.zkey build/fromAWS/${circuit_name}_vkey.json
```
Comment (motemotech) [inline circuits/scripts/download_circuits_from_aws.sh:46]: After download, add this line to automatically generate verification key

File: common/src/constants/contractConstants.ts [hunk 4]
```diff
@@ -22,3 +22,6 @@ export const PROVE_RSA_SCOPE_INDEX = 50;
 
 export const DSC_BLINDED_DSC_COMMITMENT_INDEX = 0;
 
+// Enum in VerifiersManager.sol
+export const VERIFICATION_TYPE_ENUM_PROVE = 0;
+export const VERIFICATION_TYPE_ENUM_DSC = 1;
```
Comment (motemotech) [inline common/src/constants/contractConstants.ts:27]: Add contract constant

File: common/src/utils/smtTree.ts [hunk 5]
```diff
@@ -47,6 +49,38 @@ export function buildSMT(field :any[], treetype:string): [number, number, SMT]{
...
+    const smt = new SMT(hash2, true);
+    smt.import(data.smt);
+    
+    console.log('Successfully imported SMT from JSON file');
+    return smt;
+  } catch (error) {
+      console.error('Failed to import SMT from JSON file:', error);
+      return null;
+  }
+}
```
Comment (motemotech) [inline common/src/utils/smtTree.ts:82]: add these functions to export and import smt.json.
this is used to cache smt during test for performance improvement.

File: common/src/utils/test/generateMockProof.ts
Comment (motemotech) [inline common/src/utils/test/generateMockProof.ts:1]: Changed directory name

File: contracts/README.md
Comment (motemotech) [inline contracts/README.md:1]: Add comment to run test

File: contracts/contracts/OneTimeSBT.sol
Comment (motemotech) [inline contracts/contracts/OneTimeSBT.sol:1]: Comment for future update

File: contracts/contracts/constants/Constants.sol
Comment (motemotech) [inline contracts/contracts/constants/Constants.sol:1]: Changed file name

File: contracts/contracts/verifiers/VerifiersManager.sol
Comment (motemotech) [inline contracts/contracts/verifiers/VerifiersManager.sol:1]: Comment future idea

File: contracts/package-lock.json
Comment (motemotech) [inline contracts/package-lock.json:1]: I just deleted this file while I was trying to run groth16.fullProve, so just copy again from dev branch

File: contracts/package.json
Comment (motemotech) [inline contracts/package.json:1]: add yarn command to switch environment.
import smt library
change snarkjs version

File: contracts/test/integrationTest/OneTimeVerification.ts
Comment (motemotech) [inline contracts/test/integrationTest/OneTimeVerification.ts:1]: This is the integration test for mint function in OneTimeSBT.sol

File: contracts/test/integrationTest/OneTimeVerification.ts [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+        }
+    } else {
+        prove_circuits["prove_rsa_65537_sha256"] = {
+            wasm: "../circuits/build/prove/prove_rsa_65537_sha256/prove_rsa_65537_sha256_js/prove_rsa_65537_sha256.wasm",
+            zkey: "../circuits/build/prove/prove_rsa_65537_sha256/prove_rsa_65537_sha256_final.zkey",
+            vkey: "../circuits/build/prove/prove_rsa_65537_sha256/prove_rsa_65537_sha256_vkey.json"
+        }
+    }
+
+    // TODO: After fix bug in witness gen in test code for dsc circuit
```
Comment (motemotech) [inline contracts/test/integrationTest/OneTimeVerification.ts:87]: Still I can not run groth16.fullProve in my pc.

File: contracts/test/integrationTest/OneTimeVerification.ts [hunk 7]
```diff
@@ -0,0 +1,441 @@
...
+    // TODO: I tried to generate proof in the test code, but it failed. I need to find out why
+    // TODO: export check flow in other function
+    async function generateProofDSC() {
+        let dsc = generateCircuitInputsDSC(
+            "4242",
+            mock_dsc_sha256_rsa_4096,
+            1664
+        );
+
+        // TODO: I tried to generate proof in the test code, but it failed.
```
Comment (motemotech) [inline contracts/test/integrationTest/OneTimeVerification.ts:411]: Still I can not run groth16.fullProve in my local PC

File: contracts/test/integrationTest/OneTimeVerification.ts [hunk 8]
```diff
@@ -0,0 +1,441 @@
...
+        //     vKey_csca,
+        //     publicSignals_csca,
+        //     proof_csca
+        // )
+        // // assert(verified_csca == true, 'Should verify')
+        // console.log("verified_csca: ", verified_csca);
+        // console.log('\x1b[32m%s\x1b[0m', `Proof verified csca - ${"dsc_rsa_65537_sha256_4096"}`);
+
+        const response = await sendCSCARequest(dsc);
+        console.log(response);
```
Comment (motemotech) [inline contracts/test/integrationTest/OneTimeVerification.ts:435]: So, just for now, ask modal server to get proof in integration test

File: contracts/test/integrationTest/proof_dsc.json
Comment (motemotech) [inline contracts/test/integrationTest/proof_dsc.json:1]: This is the file to be used in integration test.
Not to generate proof every time, cache the proof

File: contracts/test/integrationTest/proof_prove.json
Comment (motemotech) [inline contracts/test/integrationTest/proof_prove.json:1]: This is the file to be used in integration test.
Not to generate proof every time, cache the proof

File: contracts/test/integrationTest/smt.json
Comment (motemotech) [inline contracts/test/integrationTest/smt.json:1]: This is the file to be used in integration test.
Not to generate smt every time, cache that

File: contracts/test/unitTest/OneTimeSBT.ts
Comment (motemotech) [inline contracts/test/unitTest/OneTimeSBT.ts:1]: changed path

**Review by remicolin — APPROVED** [review]

**Pull request merged**